### PR TITLE
KAFKA-16953: Properly implement the sending of DescribeQuorumResponse

### PR DESCRIPTION
**Pull request by ahuang98**
This handles V2 of the DescribeQuorum protocol to support KIP-853

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java [hunk 1]
```diff
@@ -641,19 +641,43 @@ void assertSentDescribeQuorumResponse(
         long highWatermark,
         List<ReplicaState> voterStates,
         List<ReplicaState> observerStates
+    ) {
+        short apiVersion = (short) (kip853Rpc ? 2 : 1);
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: Take a look at `raftResponseVersion` and `describeQuorumRpcVersion`.

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java [hunk 2]
```diff
@@ -641,19 +641,43 @@ void assertSentDescribeQuorumResponse(
...
 
-        // KAFKA-16953 will add support for including the node listeners in the node collection
-        DescribeQuorumResponseData.NodeCollection nodes = new DescribeQuorumResponseData.NodeCollection();
+        if (!error.equals(Errors.NONE)) {
+            partitionData.setErrorMessage(error.message());
+        }
+
+        DescribeQuorumResponseData.NodeCollection nodes = null;
+        if (apiVersion >= 2) {
+            nodes = new DescribeQuorumResponseData.NodeCollection(voterStates.size());
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: I think the default is the empty list not the null value. For example, take a look at the implementation of `write` in `DescribeQuorumResponseData`:
```java
          if (_version >= 2) {
              _writable.writeUnsignedVarint(nodes.size() + 1);
              for (Node nodesElement : nodes) {
                  nodesElement.write(_writable, _cache, _version);
              }
          } else {
              if (!this.nodes.isEmpty()) {
                  throw new UnsupportedVersionException("Attempted to write a non-default nodes at version " + _version);
              }
          }
```

That code assumes that the value is not equals to `null`.

File: raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java [hunk 3]
```diff
@@ -183,164 +181,6 @@ public void testNonMonotonicLocalEndOffsetUpdate() {
         );
     }
 
-    @ParameterizedTest
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java:186]: Did you consider replacing these tests instead of completely removing them? Why did you decide to remove them?

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 4]
```diff
@@ -2763,78 +2765,396 @@ public void testDescribeQuorumNonLeader(boolean withKip853Rpc) throws Exception
...
+    public void testDescribeQuorumOld(boolean withKip853Rpc, short apiVersion) throws Exception {
+        int localId = 0;
+        ReplicaKey local = replicaKey(localId, withKip853Rpc);
+        ReplicaKey follower1 = replicaKey(1, withKip853Rpc);
+        Set<Integer> voters = Utils.mkSet(localId, follower1.id());
+        VoterSet voterSet = VoterSetTest.voterSet(Stream.of(local, follower1));
+
+        RaftClientTestContext.Builder builder = new RaftClientTestContext.Builder(localId, local.directoryId().orElse(ReplicaKey.NO_DIRECTORY_ID))
+            .withStaticVoters(voters)
+            .withKip853Rpc(true);
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: Why this is always set to `true`? Don't you wan to set it to `withKip853Rpc`? If you do that, I am sure you can remove the `apiVersion` parameter.

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java [hunk 5]
```diff
@@ -641,19 +641,43 @@ void assertSentDescribeQuorumResponse(
...
+        assertSentDescribeQuorumResponse(leaderId, leaderEpoch, highWatermark, voterStates, observerStates, apiVersion, Errors.NONE);
+    }
+
+    void assertSentDescribeQuorumResponse(
+        int leaderId,
+        int leaderEpoch,
+        long highWatermark,
+        List<ReplicaState> voterStates,
+        List<ReplicaState> observerStates,
+        short apiVersion,
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: I am sure you can remove this parameter if you look at my other comment about parametrizing `withKip853Rpc` and the call to the builder.

File: raft/src/main/java/org/apache/kafka/raft/LeaderState.java [hunk 6]
```diff
@@ -607,20 +565,27 @@ private void updateVoterAndObserverStates(VoterSet lastVoterSet) {
...
         long lastFetchTimestamp;
         long lastFetchLeaderLogEndOffset;
         long lastCaughtUpTimestamp;
         boolean hasAcknowledgedLeader;
 
         public ReplicaState(ReplicaKey replicaKey, boolean hasAcknowledgedLeader) {
+            this(replicaKey, hasAcknowledgedLeader, Optional.empty());
+        }
+
+        public ReplicaState(ReplicaKey replicaKey, boolean hasAcknowledgedLeader, Optional<Endpoints> listeners) {
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/LeaderState.java]: The `Endpoints` type can be empty. Did you consider using `Endpoints.empty()` instead of `Optional.empty()`?

File: raft/src/main/java/org/apache/kafka/raft/LeaderState.java [hunk 7]
```diff
@@ -301,6 +298,19 @@ public Endpoints leaderEndpoints() {
         return endpoints;
     }
 
+    int localId() {
+        return localReplicaKey.id();
+    }
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/LeaderState.java]: You can use `QuorumState.localIdOrThrow` since KRaft knows that this replica is the leader.

**Review by jsancio — COMMENTED** [review]
Thanks for the changes. I did a quick review.

File: raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java (hunk 3 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java:186]: They all are covered by new test cases I added in KafkaRaftClientTest

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java (hunk 4 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: `true` is a typo, thanks for catching that

I want this test to check the local leader is able to handle older versions of describe quorum requests (versions 0 and 1) no matter if it supports kip853Rpc or not

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java (hunk 5 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: I added this specifically for the test cases where I specify an older describeQuorumRequest version than what is default

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java (hunk 1 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java:645]: Yep, those take care of supplying the latest versions supported for describeQuorum. I wanted to add functionality to create older versions of describeQuorumResponse - in order to do that I needed to add an apiVersions param to one of these overloaded helpers and alter this one to pass in the apiVersion explicitly (which duplicates what describeQuorumRpcVersion() does). 

I'll at least use describeQuorumRpcVersion() here instead of duplicating the code

File: raft/src/main/java/org/apache/kafka/raft/LeaderState.java (hunk 6 above)
Comment (ahuang98) [reply raft/src/main/java/org/apache/kafka/raft/LeaderState.java]: I see, I missed that we will always expect `Endpoints` to be instantiated (can be empty, but not null)

File: raft/src/main/java/org/apache/kafka/raft/LeaderState.java (hunk 7 above)
Comment (ahuang98) [reply raft/src/main/java/org/apache/kafka/raft/LeaderState.java]: thanks!

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java (hunk 4 above)
Comment (jsancio) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: > I want this test to check the local leader is able to handle older versions of describe quorum requests (versions 0 and 1) no matter if it supports kip853Rpc or not

Then call `withKip853Rpc(false)`. Please take a loot at the implementation but that flag basically determines which version of the RPCs to send.

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java (hunk 5 above)
Comment (jsancio) [reply raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: See my other comment but that is what `withKip853Rpc` is for. `false` uses the old version of the RPCs. `true` uses the new version of the RPCs.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java (hunk 4 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: I'm aware that `withKip853Rpc` defines the versions of RPCs to send. The distinction I'm trying to make is `withKip853Rpc` does more than define what RPC versions to send in the test, it also defines what the local replica supports (for instance, if directory ids are supported) which can be different than what other replicas support. 

resolved offline - https://github.com/apache/kafka/pull/16637#discussion_r1689134569

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java (hunk 5 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: discussed offline, `withKip853Rpc` currently indicates more than what RPC versions the test should use (for instance if non-zero directory ids should be used in the tests), but I've convinced myself why that works. 
I'm removing `testDescribeQuorumOld` and will have the equiv functionality covered in `testDescribeQuorumWithFollowers(boolean withKip853Rpc=false)`

File: raft/src/main/java/org/apache/kafka/raft/Endpoints.java [hunk 8]
```diff
@@ -108,12 +109,25 @@ public AddRaftVoterRequestData.ListenerCollection toAddVoterRequest() {
         for (Map.Entry<ListenerName, InetSocketAddress> entry : endpoints.entrySet()) {
             listeners.add(
                 new AddRaftVoterRequestData.Listener()
-                    .setName(entry.getKey().value())
+		    .setName(entry.getKey().value())
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/Endpoints.java]: Use spaces instead of tabs.

File: raft/src/main/java/org/apache/kafka/raft/LeaderState.java [hunk 9]
```diff
@@ -647,20 +602,27 @@ private void updateVoterAndObserverStates(VoterSet lastVoterSet) {
...
         for (ReplicaState replicaStateEntry : oldVoterStates.values()) {
+            replicaStateEntry.listeners = Endpoints.empty();
             observerStates.putIfAbsent(replicaStateEntry.replicaKey, replicaStateEntry);
         }
     }
 
-    private static class ReplicaState implements Comparable<ReplicaState> {
+    static class ReplicaState implements Comparable<ReplicaState> {
         ReplicaKey replicaKey;
+        Endpoints listeners;
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/LeaderState.java]: Now that you are exposing this type to the public, let's make these fields as private and final.

File: raft/src/main/java/org/apache/kafka/raft/LeaderState.java [hunk 10]
```diff
@@ -647,20 +602,27 @@ private void updateVoterAndObserverStates(VoterSet lastVoterSet) {
...
         Optional<LogOffsetMetadata> endOffset;
         long lastFetchTimestamp;
         long lastFetchLeaderLogEndOffset;
         long lastCaughtUpTimestamp;
         boolean hasAcknowledgedLeader;
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/LeaderState.java]: Now that you are exposing this type to the public, let's make these fields as private.

File: raft/src/main/java/org/apache/kafka/raft/LeaderState.java [hunk 11]
```diff
@@ -647,20 +602,27 @@ private void updateVoterAndObserverStates(VoterSet lastVoterSet) {
...
 
         public ReplicaState(ReplicaKey replicaKey, boolean hasAcknowledgedLeader) {
+            this(replicaKey, hasAcknowledgedLeader, Endpoints.empty());
+        }
+
+        public ReplicaState(ReplicaKey replicaKey, boolean hasAcknowledgedLeader, Endpoints listeners) {
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/LeaderState.java]: Let's just have one constructor. I don't see the value of having two constructors.

File: raft/src/main/java/org/apache/kafka/raft/RaftUtil.java [hunk 12]
```diff
@@ -527,6 +541,45 @@ public static AddRaftVoterResponseData addVoterResponse(
...
+        long currentTimeMs
+    ) {
+        final long lastCaughtUpTimestamp;
+        final long lastFetchTimestamp;
+        if (replicaState.replicaKey.id() == leaderId) {
+            lastCaughtUpTimestamp = currentTimeMs;
+            lastFetchTimestamp = currentTimeMs;
+        } else {
+            lastCaughtUpTimestamp = replicaState.lastCaughtUpTimestamp;
+            lastFetchTimestamp = replicaState.lastFetchTimestamp;
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/RaftUtil.java]: These fields in ReplicaState should be private with a public accessor method. E.g.
```java
public long lastCaughtUpTimestamp() {
    return lastCaughtUpTimestamp;
}

File: raft/src/main/java/org/apache/kafka/raft/RaftUtil.java [hunk 13]
```diff
@@ -527,6 +541,45 @@ public static AddRaftVoterResponseData addVoterResponse(
...
+        }
+        DescribeQuorumResponseData.ReplicaState replicaStateData = new DescribeQuorumResponseData.ReplicaState()
+            .setReplicaId(replicaState.replicaKey.id())
+            .setLogEndOffset(replicaState.endOffset.map(LogOffsetMetadata::offset).orElse(-1L))
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/RaftUtil.java]: The `endOffset` field in ReplicaState should be private with a public accessor `public LogOffsetMetadata endOffset()`.

This comment applies to all of the fields of `ReplicaState`.

File: raft/src/main/java/org/apache/kafka/raft/internals/VoterSet.java [hunk 14]
```diff
@@ -321,7 +321,7 @@ public boolean isVoter(ReplicaKey replicaKey) {
             }
         }
 
-        Endpoints listeners() {
+        public Endpoints listeners() {
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/internals/VoterSet.java]: Add a java doc section.

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java [hunk 15]
```diff
@@ -648,19 +648,41 @@ void assertSentDescribeQuorumResponse(
...
+        assertSentDescribeQuorumResponse(leaderId, leaderEpoch, highWatermark, voterStates, observerStates, Errors.NONE);
+    }
+
+    void assertSentDescribeQuorumResponse(
+        int leaderId,
+        int leaderEpoch,
+        long highWatermark,
+        List<ReplicaState> voterStates,
+        List<ReplicaState> observerStates,
+        Errors error
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: The order of this parameter doesn't match the order used in other methods. `Errors` parameter tends to go first in the method parameter list.

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java [hunk 16]
```diff
@@ -648,19 +648,41 @@ void assertSentDescribeQuorumResponse(
...
     ) {
         DescribeQuorumResponseData response = collectDescribeQuorumResponse();
 
         DescribeQuorumResponseData.PartitionData partitionData = new DescribeQuorumResponseData.PartitionData()
-            .setErrorCode(Errors.NONE.code())
+            .setErrorCode(error.code())
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java:668]: This comment applies to the entire method. Why reimplement this instead of using `RaftUtil.singletonDescribeQuorumResponse`?

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 17]
```diff
@@ -2762,79 +2764,357 @@ public void testDescribeQuorumNonLeader(boolean withKip853Rpc) throws Exception
...
+    @Test
+    public void testDescribeQuorumWithOnlyStaticVoters() throws Exception {
+        int localId = 0;
+        ReplicaKey local = replicaKey(localId, true);
+        ReplicaKey follower1 = replicaKey(1, true);
+        Set<Integer> voters = Utils.mkSet(localId, follower1.id());
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(localId, local.directoryId().get())
+            .withStaticVoters(voters)
+            .withKip853Rpc(true)
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: Should we test this case with both the new and old RPC versions?

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 18]
```diff
@@ -2762,79 +2764,357 @@ public void testDescribeQuorumNonLeader(boolean withKip853Rpc) throws Exception
         assertEquals(Errors.NOT_LEADER_OR_FOLLOWER.message(), partitionData.errorMessage());
     }
 
+    @Test
+    public void testDescribeQuorumWithOnlyStaticVoters() throws Exception {
+        int localId = 0;
+        ReplicaKey local = replicaKey(localId, true);
+        ReplicaKey follower1 = replicaKey(1, true);
+        Set<Integer> voters = Utils.mkSet(localId, follower1.id());
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: Let's make the id randoms. Please take a look at other methods/tests for an example of how to do this.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 19]
```diff
@@ -2762,79 +2764,357 @@ public void testDescribeQuorumNonLeader(boolean withKip853Rpc) throws Exception
...
-        RaftClientTestContext context = new RaftClientTestContext.Builder(localId, voters)
-            .withKip853Rpc(withKip853Rpc)
-            .build();
+    public void testDescribeQuorumWithFollowers(boolean withKip853Rpc) throws Exception {
+        int localId = 0;
+        ReplicaKey local = replicaKey(localId, true);
+        Uuid localDirectoryId = local.directoryId().get();
+        ReplicaKey follower1 = replicaKey(1, true);
+        Uuid followerDirectoryId1 = follower1.directoryId().get();
+        ReplicaKey follower2 = replicaKey(2, false);
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: Let's make the replica ids random. Please take a look at other methods/tests for an example on how to do this.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 20]
```diff
@@ -2762,79 +2764,357 @@ public void testDescribeQuorumNonLeader(boolean withKip853Rpc) throws Exception
...
+        Set<Integer> voters = Utils.mkSet(localId, follower1.id(), follower2.id());
+        VoterSet voterSet = VoterSetTest.voterSet(Stream.of(local, follower1, follower2));
+
+        RaftClientTestContext.Builder builder = new RaftClientTestContext.Builder(localId, localDirectoryId)
+            .withStaticVoters(voters)
+            .withKip853Rpc(withKip853Rpc);
+
+        if (withKip853Rpc) {
+            builder.withBootstrapSnapshot(Optional.of(voterSet));
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: Did you meant to set both the static voters and the bootstrap checkpoint when `withKip853Rpc` is true?

This comment applies to a few places.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 21]
```diff
@@ -2762,79 +2764,357 @@ public void testDescribeQuorumNonLeader(boolean withKip853Rpc) throws Exception
...
+        context.pollUntilResponse();
 
+        expectedVoterStates.get(0)
+            .setLogEndOffset(nextFetchOffset)
+            .setLastFetchTimestamp(context.time.milliseconds())
+            .setLastCaughtUpTimestamp(context.time.milliseconds());
+        expectedVoterStates.get(1)
+            .setLogEndOffset(fetchOffset)
+            .setLastFetchTimestamp(followerFetchTime1)
+            .setLastCaughtUpTimestamp(followerFetchTime1);
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java:2883]: `followerFetchTime1` is `context.time.milliseconds` since you didn't sleep after sending the FETCH request. In other words, you can remove the variable `followerFetchTime1` and make it easier to read.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 22]
```diff
@@ -2762,79 +2764,357 @@ public void testDescribeQuorumNonLeader(boolean withKip853Rpc) throws Exception
...
+            .setLastCaughtUpTimestamp(followerFetchTime2);
+        context.assertSentDescribeQuorumResponse(localId, epoch, expectedHW, expectedVoterStates, Collections.emptyList());
+
+        // Describe quorum returns error if leader loses leadership
+        context.time.sleep(context.checkQuorumTimeoutMs);
+        context.deliverRequest(context.describeQuorumRequest());
+        context.pollUntilResponse();
+        expectedVoterStates.get(0)
+            .setLastFetchTimestamp(context.time.milliseconds())
+            .setLastCaughtUpTimestamp(context.time.milliseconds());
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: Why is this necessary? The object `expectedVoterStates` is not used after this point.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 23]
```diff
@@ -2762,79 +2764,357 @@ public void testDescribeQuorumNonLeader(boolean withKip853Rpc) throws Exception
...
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = { true, false })
+    public void testDescribeQuorumWithObserver(boolean withKip853Rpc) throws Exception {
+        int localId = 0;
+        ReplicaKey local = replicaKey(localId, true);
+        Uuid localDirectoryId = local.directoryId().get();
+        ReplicaKey follower = replicaKey(1, withKip853Rpc);
+        Uuid followerDirectoryId = withKip853Rpc ? follower.directoryId().get() : ReplicaKey.NO_DIRECTORY_ID;
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: You can use Optional's orElse.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java (hunk 19 above)
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: Looks like you want to use `replicaKey(someId, withKip853Rpc)` and compute the directory id sent over the RPC with `replicaKey.directoryId().orElse(ReplicaKey.NO_DIRECTORY_ID)`.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 24]
```diff
@@ -2762,79 +2764,357 @@ public void testDescribeQuorumNonLeader(boolean withKip853Rpc) throws Exception
...
+        context.pollUntilResponse();
+
+        expectedVoterStates.get(0)
+            .setLastFetchTimestamp(context.time.milliseconds())
+            .setLastCaughtUpTimestamp(context.time.milliseconds());
+        context.assertSentDescribeQuorumResponse(localId, epoch, expectedHW, expectedVoterStates, Collections.emptyList());
+    }
+
+    @Test
+    public void testDescribeQuorumNonMonotonicFollowerFetch() throws Exception {
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: For some of these tests it looks like you want the following CSV source:
```
withKip853Rpc,withBootstrapCheckpoint
false,false
true,true
true,false
```

**Review by jsancio — COMMENTED** [review]
Thanks for the changes.

File: raft/src/main/java/org/apache/kafka/raft/Endpoints.java (hunk 8 above)
Comment (ahuang98) [reply raft/src/main/java/org/apache/kafka/raft/Endpoints.java]: thanks, wonky settings from my new laptop

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java (hunk 16 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java:668]: Makes sense to have less duplicate code. I'll change it to use RaftUtil.singletonDescribeQuorumResponse. The small difference is assertSentDescribeQuorumResponse takes the voter and observer states as is while RaftUtil.singletonDescribeQuorumResponse will accept an additional `currentTimeMs` and overwrite the leader's `lastCaughtUp` and `lastFetch` timestamps with it

Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java:668]: Actually, `assertSentDescribeQuorumResponse` (and the test callers) use DescribeQuorumResponseData.ReplicaState while RaftUtil uses LeaderState.ReplicaState, which is a bit annoying but I can make the conversion

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java (hunk 20 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: My reasoning was that it shouldn't break to set both (we'll ignore static when we have a bootstrap checkpoint), but I can be more explicit about it - i.e. make it either or for all existing tests and add `testStaticVotersIgnoredWithBootstrapSnapshot`

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java (hunk 21 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java:2883]: opted for adding a sleep

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java (hunk 20 above)
Comment (jsancio) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: Yes. If you want to test that case it should be an explicit tests. Having said that, supporting both static voter and bootstrap voters has nothing to do with the DescribeQuorum RPC.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 25]
```diff
@@ -2764,77 +2766,414 @@ public void testDescribeQuorumNonLeader(boolean withKip853Rpc) throws Exception
...
+            new ReplicaState()
+                .setReplicaId(follower2.id())
+                .setReplicaDirectoryId(withKip853Rpc ? follower2.directoryId().get() : ReplicaKey.NO_DIRECTORY_ID)
+                .setLogEndOffset(-1L)
+                .setLastFetchTimestamp(-1)
+                .setLastCaughtUpTimestamp(-1));
+        context.assertSentDescribeQuorumResponse(localId, epoch, -1L, expectedVoterStates, Collections.emptyList());
+    }
+
+    // After KAFKA-16535 we can test describe quorum output following voter removal and addition
```
Comment (ahuang98) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: will remove this comment.

**Comment by ahuang98** [comment]
Looks like another tab character snuck in. Going to push a commit to fix the build

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 26]
```diff
@@ -2803,21 +2803,28 @@ public void testDescribeQuorumWithOnlyStaticVoters(boolean withKip853Rpc) throws
...
+        Uuid localDirectoryId = local.directoryId().orElse(Uuid.randomUuid());
+        ReplicaKey bootstrapFollower1 = replicaKey(followerId1, withBootstrapSnapshot);
+        // if withBootstrapSnapshot is false, directory ids are still needed by the static voter set
+        Uuid followerDirectoryId1 = bootstrapFollower1.directoryId().orElse(withKip853Rpc ? Uuid.randomUuid() : ReplicaKey.NO_DIRECTORY_ID);
+        ReplicaKey follower1 = ReplicaKey.of(followerId1, followerDirectoryId1);
+        ReplicaKey bootstrapFollower2 = replicaKey(followerId2, withBootstrapSnapshot);
+        Uuid followerDirectoryId2 = bootstrapFollower2.directoryId().orElse(withKip853Rpc ? Uuid.randomUuid() : ReplicaKey.NO_DIRECTORY_ID);
+        ReplicaKey follower2 = ReplicaKey.of(followerId2, followerDirectoryId2);
+        VoterSet bootstrapVoterSet = VoterSetTest.voterSet(Stream.of(local, bootstrapFollower1, bootstrapFollower2));
+        VoterSet staticVoterSet = VoterSetTest.voterSet(Stream.of(local, follower1, follower2));
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: You can move this to the `if (withBootstrapSnapshot) ... else ...` state and avoid creating too many local variable that are not used through out the method.

This comment applies to a few places.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java (hunk 25 above)
Comment (jsancio) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: TODO reminder.

**Review by jsancio — COMMENTED** [review]
Thanks for the changes. We are close just some minor comments.

**Comment by jsancio** [comment]
@ahuang98 There are failures in the MetadataQuorumCommandTest suite.

File: raft/src/main/java/org/apache/kafka/raft/LeaderState.java [hunk 27]
```diff
@@ -702,27 +657,50 @@ private void updateVoterAndObserverStates(VoterSet lastVoterSet) {
...
     }
 
-    private static class ReplicaState implements Comparable<ReplicaState> {
-        ReplicaKey replicaKey;
-        Optional<LogOffsetMetadata> endOffset;
-        long lastFetchTimestamp;
-        long lastFetchLeaderLogEndOffset;
-        long lastCaughtUpTimestamp;
-        boolean hasAcknowledgedLeader;
+    static class ReplicaState implements Comparable<ReplicaState> {
```
Comment (jlprat) [inline raft/src/main/java/org/apache/kafka/raft/LeaderState.java:665]: This static class should be public as it is exposed in `RaftUtil.singletonDescribeQuorumResponse`. Right now this class is exposed outside of it's visibility scope.
